Add Salvium (SAL) Integration

whiskyrelaxing opened the pull request:
Overview:
Adds full support for Salvium cryptocurrency to BasicSwapDEX, including the new Carrot addressing format (SC1 prefix) introduced in Salvium One hard fork.

Changes:
- Chainparams: Added SAL with Carrot address prefixes (0x180c96 mainnet, 0x254c96 testnet, 0x24cc96 regtest)
- Interface: Created sal.py extending XMRInterface with Salvium-specific methods
  - Carrot address derivation (getAddressFromKeys, encodeSharedAddress)
  - SAL1 asset type handling in wallet operations
  - Override getSpendableBalance() for SAL's balances array structure
  - Graceful handling of empty wallets
- UI: Updated templates to include SAL alongside XMR/WOW
  - Wallet page with deposit/withdraw sections
  - Added |int filters to template coin ID comparisons
  - Offers page coin listing
- JavaScript: Added SAL ('19') to adaptor_sig_only_coins arrays
  - Price tracking, coin management, and charts
- Prepare/Run Scripts: Added SAL daemon/wallet download and configuration

Tested:
- Deposits/withdrawals with SC1 Carrot addresses
- Fee estimation
- Swap offer creation and revocation (SAL ↔ PART)

Requirements:
- Requires Salvium 1.0.6+ with Carrot addressing support
- Use --withcoins=salvium in basicswap-prepare
- Clean salvium wallet recommended (no migration from pre-Carrot wallets)

Commit:
Add Salvium (SAL) Integration

Overview:
Adds full support for Salvium cryptocurrency to BasicSwapDEX, including the new Carrot addressing format (SC1 prefix) introduced in Salvium One hard fork.

Changes:
- Chainparams: Added SAL with Carrot address prefixes (0x180c96 mainnet, 0x254c96 testnet, 0x24cc96 regtest)
- Interface: Created sal.py extending XMRInterface with Salvium-specific methods
  - Carrot address derivation (getAddressFromKeys, encodeSharedAddress)
  - SAL1 asset type handling in wallet operations
  - Override getSpendableBalance() for SAL's balances array structure
  - Graceful handling of empty wallets
- UI: Updated templates to include SAL alongside XMR/WOW
  - Wallet page with deposit/withdraw sections
  - Added |int filters to template coin ID comparisons
  - Offers page coin listing
- JavaScript: Added SAL ('19') to adaptor_sig_only_coins arrays
  - Price tracking, coin management, and charts
- Prepare/Run Scripts: Added SAL daemon/wallet download and configuration

Tested:
- Deposits/withdrawals with SC1 Carrot addresses
- Fee estimation
- Swap offer creation and revocation (SAL ↔ PART)

Requirements:
- Requires Salvium 1.0.6+ with Carrot addressing support
- Use --withcoins=salvium in basicswap-prepare
- Clean salvium wallet recommended (no migration from pre-Carrot wallets)

nahuhh commented:
This has to be rebased on the `dev` branch

Commit:
Fix WOW chain state polling - use threadPollXMRChainState

nahuhh commented on basicswap/basicswap.py:
```diff
@@ -254,6 +254,25 @@ def threadPollXMRChainState(swap_client, coin_type):
+        swap_client.chainstate_delay_event.wait( 
+            random.randrange(20, 30)
+        )  # Random to stagger updates
 
```
The threadpollxmrchainstate can be reused. deleted

nahuhh commented on basicswap/basicswap.py:
```diff
@@ -1173,6 +1197,7 @@ def start(self):
 
                 thread_func = {
                     Coins.XMR: threadPollXMRChainState,
+                    Coins.SAL: threadPollSALChainState,
```
```suggestion
                    Coins.SAL: threadPollXMRChainState,
```

Can probably be cleaned up to smthn like
```
if c in (Coins.XMR, Coins.SAL, Coins.WOW):
    thread_func = threadPollXMRChainstate
else:
    thread_func = threadPollChainstate
```

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -359,6 +374,36 @@ def setTorrcVars():
+wownero_wallet_rpc_proxy_config = [
+    #   'daemon-ssl-allow-any-cert=1', moved to startup flag
+]
+
```
Botched rebase - delete

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -359,6 +374,36 @@ def setTorrcVars():
+salvium_wallet_rpc_proxy_config = [
+    #   'daemon-ssl-allow-any-cert=1', moved to startup flag
+]
+
```
Delete

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -359,6 +374,36 @@ def setTorrcVars():
+    "out-peers=24",
+    f"tx-proxy=tor,{TOR_PROXY_HOST}:{TOR_PROXY_PORT},disable_noise,16",  # Outgoing tx relay to onion
+]
+
```
Does salvium have onion seed nodes? And a healthy onion network? (Has peers?)

If yes, this can be deleted.

If no, you'll have to edit this to remove tx-proxy

update: salvium does _not_, and irresponsibly has monero's onion (and i2p) seed nodes still hardcoded.

```suggestion
salviumd_proxy_config = [
    f"proxy={TOR_PROXY_HOST}:{TOR_PROXY_PORT}",
    "proxy-allow-dns-leaks=0",
    "no-igd=1",  # Disable UPnP port mapping
    "hide-my-port=1",  # Don't share the p2p port
    "p2p-bind-ip=127.0.0.1",  # Don't broadcast ip
    "in-peers=0",  # Changes "error" in log to "incoming connections disabled"
    "out-peers=24"
]

```

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -651,36 +696,52 @@ def extractCore(coin, version_data, settings, bin_dir, release_path, extra_opts=
             raise ValueError("Unknown coin")
 
-        if "win32" in BIN_ARCH or "win64" in BIN_ARCH:
+        # Salvium uses zip files, others use tar
```
Salvium should get w the program and use tar for the non-windows releases

whiskyrelaxing replied:
I am not a member of the salvium team and have no control over their releases. I was mirroring their release setup and using their release tools to build the ubuntu zip file I was hosting for this but I can make it a .tar.gz file instead.

nahuhh replied:
You cant make it a tar - the files have to be signed releases with trusted keys

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -651,36 +696,52 @@ def extractCore(coin, version_data, settings, bin_dir, release_path, extra_opts=
                                 break
             return
+        else:
+            if "win32" in BIN_ARCH or "win64" in BIN_ARCH:
```
Use elif

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -815,6 +876,33 @@ def prepareCore(coin, version_data, settings, data_dir, extra_opts={}):
+            else:
+                platform_name = "macos-x86_64"
+            release_url = f"https://github.com/salvium/salvium/releases/download/v{version}/salvium-v{version}-{platform_name}.{use_file_ext}"
+        else:  # Linux - use Ubuntu 22.04 build for BasicSwap compatibility
```
No support for aarch64 linux?

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -815,6 +876,33 @@ def prepareCore(coin, version_data, settings, data_dir, extra_opts={}):
+        downloadRelease(release_url, release_path, extra_opts)
+
+        # Skip verification and extract
+        extractCore(coin, version_data, settings, bin_dir, release_path, extra_opts)
```
Not a chance

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1216,11 +1304,22 @@ def prepareDataDir(coin, settings, chain, particl_mnemonic, extra_opts={}):
+                if tor_control_password is not None:
+                    for opt_line in monerod_proxy_config:
+                        fp.write(opt_line + "\n")
+
```
Botched rebase. Delete

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1216,11 +1304,22 @@ def prepareDataDir(coin, settings, chain, particl_mnemonic, extra_opts={}):
+                if tor_control_password is not None:
+                    for opt_line in salviumd_proxy_config:
+                        fp.write(opt_line + "\n")
+
```
if salvium has healthy/viable onion peers and seeds, delete

update: they do not

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1216,11 +1304,22 @@ def prepareDataDir(coin, settings, chain, particl_mnemonic, extra_opts={}):
                     fp.write(f"rpc-login={WOW_RPC_USER}:{WOW_RPC_PWD}\n")
             if tor_control_password is not None:
-                for opt_line in monerod_proxy_config:
+                for opt_line in wownerod_proxy_config:
```
Botched rebase 

```suggestion
                for opt_line in monerod_proxy_config:
```

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1260,6 +1359,12 @@ def prepareDataDir(coin, settings, chain, particl_mnemonic, extra_opts={}):
+                    "shared-ringdb-dir={}\n".format(
+                        os.path.join(config_datadir, "shared-ringdb")
+                    )
+                )
```
Deduplicate.

Change line 1356 to

elif coin in ("monero", "salvium")

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1876,6 +1986,21 @@ def initialise_wallets(
+                            coin_settings["bindir"],
+                            filename,
+                        )
+                    )
```
Deduplicate

line 1974 to 

`if c in (Coins.XMR, Coins.SAL):`

nahuhh commented on basicswap/chainparams.py:
```diff
@@ -12,7 +12,7 @@
 XMR_COIN = 10**12
 WOW_COIN = 10**11
-
+SAL_COIN = 10**8
```
Missing blank line after this

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -2981,6 +3145,25 @@ def main():
+                prepareDataDir(
+                    add_coin, settings, chain, particl_wallet_mnemonic, extra_opts
+                )
+
```
Botched rebase

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -2955,6 +3113,12 @@ def main():
     extra_opts["data_dir"] = data_dir
     extra_opts["tor_control_password"] = tor_control_password
 
+    # Debug: Check if chainclients is properly populated
+    if "salvium" in chainclients:
+        logger.info(f"Salvium chainclients keys before addcoin: {list(chainclients['salvium'].keys())}")
+        if "walletrpcport" in chainclients["salvium"]:
+            logger.info(f"walletrpcport value: {chainclients['salvium']['walletrpcport']}")
+
```
?

nahuhh commented on Dockerfile:
```diff
@@ -6,7 +6,7 @@ ENV LANG=C.UTF-8 \
 RUN apt-get update; \
     apt-get install -y --no-install-recommends \
-        python3-pip libpython3-dev gnupg pkg-config gcc libc-dev gosu tzdata cmake ninja-build;
+        python3-pip libpython3-dev gnupg pkg-config gcc libc-dev gosu tzdata cmake ninja-build curl sqlite3 wget bzip2;
```
Why are these added

Commit:
Address PR review items
- Remove duplicate threadPollSALChainState, use threadPollXMRChainState
- Add checksum verification with hosted hashes file
- Remove Tor proxy config for salvium
- Deduplicate monero/salvium code sections
- Remove debug logging
- Fix package list
- Clean up rebase artifacts

Commit:
Fix missed case: threadPollChainstate -> threadPollChainState

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -359,6 +374,7 @@ def setTorrcVars():
     #   'daemon-ssl-allow-any-cert=1', moved to startup flag
 ]
 
+
```
You'll want a salviumd_proxy_config here, but without the tx-proxy

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1216,10 +1276,15 @@ def prepareDataDir(coin, settings, chain, particl_mnemonic, extra_opts={}):
-            if tor_control_password is not None:
+
+            # Tor proxy for monero/wownero only
+            if coin in ("monero", "wownero") and tor_control_password is not None:
```
You'll want one of these w/ salviumd_proxy_config

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -2981,6 +3090,20 @@ def main():
+                prepareDataDir(
+                    add_coin, settings, chain, particl_wallet_mnemonic, extra_opts
+                )
+
```
Rm

nahuhh commented on basicswap/interface/sal.py:
```diff
@@ -0,0 +1,496 @@
+import basicswap.util_xmr as xmr_util
+from basicswap.chainparams import SAL_COIN, Coins
+from basicswap.util import b2h, TemporaryError
+from .xmr import XMRInterface
+
+
+class SALInterface(XMRInterface):
```
Since sal isnt based on the latest monero release, it should probably use the same overrides as wownero

Commit:
address additional PR items
- Add salviumd_proxy_config with basic Tor support (no tx-proxy)
- Include salvium in Tor proxy conditional
- Add openWallet override to SALInterface matching wownero
note: many SAL rpc calls require the use of asset type
- Remove rebase artifact

nahuhh commented on basicswap/basicswap.py:
```diff
@@ -228,7 +228,6 @@ def checkAndNotifyBalanceChange(
         cc["cached_total_balance"] = None
         cc["cached_unconfirmed"] = None
 
-
```
Undo whitespace change

nahuhh commented on basicswap/basicswap.py:
```diff
@@ -254,7 +253,6 @@ def threadPollXMRChainState(swap_client, coin_type):
             random.randrange(20, 30)
         )  # Random to stagger updates
 
-
```
Undo whitespace change

nahuhh commented on basicswap/basicswap.py:
```diff
@@ -286,7 +284,6 @@ def threadPollChainState(swap_client, coin_type):
             swap_client.log.warning(f"threadPollChainState {ci.ticker()}, error: {e}")
         swap_client.chainstate_delay_event.wait(random.randrange(*poll_delay_range))
 
-
```
Undo whitespace change

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -651,36 +676,52 @@ def extractCore(coin, version_data, settings, bin_dir, release_path, extra_opts=
         else:
             raise ValueError("Unknown coin")
 
-        if "win32" in BIN_ARCH or "win64" in BIN_ARCH:
```
Why not just add `or coin == "salvium"`

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -1216,12 +1285,21 @@ def prepareDataDir(coin, settings, chain, particl_mnemonic, extra_opts={}):
-                    fp.write(opt_line + "\n")
+
+            # Tor proxy for salvium or monero/wownero only
+            if coin in ("monero", "wownero", "salvium") and tor_control_password is not None:
```
L1260 already gates this to these coins

```suggestion
            if tor_control_password is not None:
```

nahuhh commented on basicswap/bin/prepare.py:
```diff
@@ -815,6 +856,34 @@ def prepareCore(coin, version_data, settings, data_dir, extra_opts={}):
         assert_path = os.path.join(bin_dir, assert_filename)
         if not os.path.exists(assert_path):
             downloadFile(assert_url, assert_path)
+
+    elif coin == "salvium":
+        use_file_ext = "zip"
+    
+        if "win32" in BIN_ARCH or "win64" in BIN_ARCH:
+            platform_name = "win64"
+            release_url = f"https://github.com/salvium/salvium/releases/download/v{version}/salvium-v{version}-{platform_name}.{use_file_ext}"
+        elif "osx" in BIN_ARCH:
+            if "arm" in BIN_ARCH or "aarch64" in BIN_ARCH:
+                platform_name = "macos-arm64"
+            else:
+                platform_name = "macos-x86_64"
+            release_url = f"https://github.com/salvium/salvium/releases/download/v{version}/salvium-v{version}-{platform_name}.{use_file_ext}"
+        else:  # Linux
+            platform_name = "ubuntu22.04-linux-x86_64"
+            use_file_ext = "tar.gz"
+            release_url = f"https://www.whiskymine.io/salvium-v{version}-{platform_name}.{use_file_ext}"
+    
+        release_filename = f"salvium-v{version}-{platform_name}.{use_file_ext}"
+        release_path = os.path.join(bin_dir, release_filename)
+        downloadRelease(release_url, release_path, extra_opts)
+    
+        assert_filename = f"salvium-{version}-hashes.txt"
+        assert_url = f"https://www.whiskymine.io/salvium-{version}-hashes.txt"
+        assert_path = os.path.join(bin_dir, assert_filename)
+        if not os.path.exists(assert_path):
+            downloadFile(assert_url, assert_path)
+
```
This is a blocker and will have to be redone. Can only accept official release binaries, signed by the developers (and they will have to release binaries for all platforms)

Commit:
Address PR feedback items 1-3

- Restore 2 blank lines between functions (file style consistency)
- Simplify extractCore: salvium uses same zip logic as Windows
- Simplify Tor proxy: remove redundant coin check (already gated at parent level)

Note: Item 4 (official signed binaries) pending Salvium team response

nahuhh commented:
drop the merge commit please. `git reset --hard HEAD~1`

rebase onto **dev** if you need to pull in updates `git fetch origin && git rebase origin/dev`

Commit:
Update Salvium to v1.0.7 with official GitHub releases, Bump version from 1.0.6 to 1.0.7, Add proper hash verification, Lock hash file to commit

whiskyrelaxing commented:
Following up, the last push did have a git reset --hard HEAD~1 and fetch/rebase as requested. Is there anything else required?

whiskyrelaxing commented:
Rebased onto latest dev and resolved conflicts

nahuhh commented:
You've got some linting issues (flake8)

You can also use pre-commit hooks:
1. Install pre-commit package: `pip install pre-commit`
2. Install hooks to the repo: `pre-commit install`

This should check your linting when you make commits to this repo

Commit:
Fix linting issues for Salvium PR, ran flake8 and black tools, integrated with latest dev

Commit:
Add fast-block-sync=0 to salviumd config
